File: 1127-last-stone-weight/1127-last-stone-weight.js
/**
 * @param {number[]} stones
 * @return {number}
 */
var lastStoneWeight = function (stones) {
    let heap = new MaxPriorityQueue()
    for (let stone of stones) { 
        heap.enqueue(stone, stone); 
    }
    let arr = []
    while (heap.size() > 1) {
        let first = heap.dequeue().element
        let second = heap.dequeue().element


        if (first !== second) {
            heap.enqueue(first - second, first - second)
        }
    }
    return heap.isEmpty() ? 0 : heap.dequeue().element
};

File: 1127-last-stone-weight/1127-last-stone-weight.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        maxHeap = [-stone for stone in stones]
        heapq.heapify(maxHeap)

        
        while(len(maxHeap) > 1):
            first = -heapq.heappop(maxHeap)
            second = -heapq.heappop(maxHeap)
            if first != second:
                heapq.heappush(maxHeap, -(first - second))

        return -maxHeap[0] if maxHeap else 0
